### PR TITLE
Full support for printing and parsing transactions as JSON

### PR DESCRIPTION
**Pull request by ggrieco-tob**
This PR is the first step to allow Echidna to save and load list of transaction as JSON.

### CODE REVIEW

**Comment by incertia** [comment]
can we get some tests

**Comment by ggrieco-tob** [comment]
@incertia @arcz I just added the test. Feel free to take a look while I fix the code climate issues.

File: src/test/Spec.hs [hunk 1]
```diff
@@ -257,3 +262,39 @@ solvedLen i t = (== Just i) . fmap length . solnFor t
...
+instance Arbitrary Tx where
+  arbitrary = do
+               t <- arbitrary :: Gen TxCall
+               s <- arbitrary :: Gen Addr
+               d <- arbitrary :: Gen Addr
+               g <- arbitrary :: Gen EVM.Concrete.Word
+               gp <- arbitrary :: Gen EVM.Concrete.Word
+               v <- arbitrary :: Gen EVM.Concrete.Word
+               dl <- arbitrary :: Gen (EVM.Concrete.Word, EVM.Concrete.Word)
+               return $ Tx t s d g gp v dl
```
Comment (incertia) [inline src/test/Spec.hs]: we can shorten this to
```haskell
let a = arbitrary
Tx <$> a <*> a <*> ... <*> a
```

File: src/test/Spec.hs [hunk 2]
```diff
@@ -257,3 +262,39 @@ solvedLen i t = (== Just i) . fmap length . solnFor t
...
+
+instance Arbitrary Addr where
+  arbitrary = arbitrary >>= (return . fromInteger)
+
+instance Arbitrary EVM.Concrete.Word where
+  arbitrary = arbitrary >>= (return . fromInteger)
+
+instance Arbitrary TxCall where
+  arbitrary = do
+              s <- arbitrary :: Gen String
```
Comment (incertia) [inline src/test/Spec.hs]: indent 2 spaces on `do` and not to the `do`

File: lib/Echidna/Transaction.hs [hunk 3]
```diff
@@ -59,6 +63,44 @@ data Tx = Tx { _call  :: TxCall       -- | Call
...
+instance ToJSON Addr where
+  toJSON = toJSON . show
+
+instance ToJSON Word where
+  toJSON = toJSON . show
+
+instance FromJSON Word where
+  parseJSON = withText "Word" $ maybe (M.fail "could not parse Word") pure . readMaybe . T.unpack
+
+$(deriveJSON defaultOptions ''EVM.ABI.AbiType)
```
Comment (incertia) [inline lib/Echidna/Transaction.hs]: do we really need `EVM.ABI` here?

File: src/test/Spec.hs [hunk 4]
```diff
@@ -257,3 +262,39 @@ solvedLen i t = (== Just i) . fmap length . solnFor t
...
+              cs <- arbitrary :: Gen [AbiValue]
+              return $ SolCall (pack s, cs)
```
Comment (incertia) [inline src/test/Spec.hs]: the typechecker should figure out the types of `arbitrary`

File: src/test/Spec.hs [hunk 5]
```diff
@@ -257,3 +262,39 @@ solvedLen i t = (== Just i) . fmap length . solnFor t
...
+
+encodingJSONTests :: TestTree
+encodingJSONTests =
+  testGroup "Tx JSON encoding"
+    [ testProperty "decode . encode = id" $ property $ do
+        t <- arbitrary :: Gen Tx
+        let jt = encode t
+        let t' = decode jt :: Maybe Tx
+        return $ t' === Just t
+    ]
```
Comment (incertia) [inline src/test/Spec.hs]: maybe
```haskell
return $ decode (encode t) === Just t
```

File: src/test/Spec.hs [hunk 6]
```diff
@@ -257,3 +262,39 @@ solvedLen i t = (== Just i) . fmap length . solnFor t
...
+               dl <- arbitrary :: Gen (EVM.Concrete.Word, EVM.Concrete.Word)
+               return $ Tx t s d g gp v dl
+
+encodingJSONTests :: TestTree
+encodingJSONTests =
+  testGroup "Tx JSON encoding"
+    [ testProperty "decode . encode = id" $ property $ do
+        t <- arbitrary :: Gen Tx
+        let jt = encode t
+        let t' = decode jt :: Maybe Tx
```
Comment (incertia) [inline src/test/Spec.hs]: if `t` is `Tx` then this is already `Maybe Tx`
